eng/pipelines/evaluate-changed-paths.sh

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env bash
+: '
+Scenarios:
+  1. exclude paths are specified
+      Will include all paths except the ones in the exclude list.
+  2. include paths are specified
+      Will only include paths specified in the list.
+  3. exclude + include:
+      1st we evaluate changes for all paths except ones in excluded list. If we can not find
+      any applicable changes like that, then we evaluate changes for incldued paths
+      if any of these two finds changes, then a variable will be set to true.
+  In order to consume this variable you need to reference it via: $[ dependencies.checkout.outputs["SetPathVars_<subset>.containschange"] ]
+
+  Example:
+  -difftarget ''HEAD^1'' -subset coreclr -includepaths ''src/libraries/System.Private.CoreLib/*'' -excludepaths ''src/libraries/*+src/installer/*''
+
+  This example will include ALL path changes except the ones under src/libraries/*!System.Private.CoreLib/*
+'
+
+# Disable globbing in this bash script since we iterate over path patterns
+set -f
```

Comment:
Definitely agree. In fact this should be our default for bash script writing. 

Do we have guidelines for how we write bash and powershell scripts? Example document for Powershell that we use. Essentially all the best practices, aka lessons learned from mistakes, that we want to use going forward. 

https://github.com/dotnet/roslyn/blob/master/docs/contributing/Powershell%20Guidelines.md

Reply:
The reaso why I didn’t use -e is because since I’m relying on git diff exit code to know if a path was touched the script will exit if that’s true, git diff will return 1 when changes are present and 0 when no changes are present.

Is there a workaround to that?

Reply:
Spoke offline. @dagood explained how to do that. Will update.

Reply:
So how do you do that? Cause I want to know to. 

Might make an interesting addition to a "Bash Guidelines.md" document 😄

Reply:
My answer in the chat was basically "don't", as far as using `$?` goes in this case. Instead, let the exit code flow out of the function. Outside, use the function in an `if` so it doesn't terminate the script as a whole. 

(Related: the suggestion later to use `if ! runGitDiff $exclude_path_string; then`.)

Two other alternatives if you really need to store it in a variable (didn't think of these at the time):
* `thingthatwillfail || x=$?`. The exit code doesn't make it to the script, and `x=$?` succeeds, so it keeps going. `x` is unset rather than `0` this way, but that can be accounted for.
* `(set +e; thingthatwillfail; x=$? )`. `set +e` reverses `set -e`, the subshell makes sure it returns to whatever it was before, and the last statement to run is `x=$?` which exits successfully, so the subshell as a whole has a success error code. (You may want to replace `;` with newlines, especially if it's a longer section you want to run with `set +e`.)